doc/spec/namespace.md

```diff
@@ -0,0 +1,149 @@
+- *Name* - A fully qualified string containing both the domain and resource name
+- *Repository* - A collection of objects under the same name within the
+namespace.
+- *Namespace* - also *Namespace Scope* - A collection of repositories with a
+common name prefix and set of services including registry api, index, and trust
+context.
+- *Short Name* - A name which does not contain a domain and requires
+expansion to a fully qualified name before resolving to a repository.
+
+## Format
+A name consists of two parts, a DNS host name plus a repository path. The host
+name follows the DNS rules without change. The total length of a name is 255
+characters however there is no specific limitation on DNS or path components.
+
+### Name Grammar
+~~~
+<name> ::= <hostname>"/"<path>
+<hostname> ::= <host-part>*["."<host-part>][":"<port>]
+<path> ::= <path-part>*["/"<path-part>]
+<host-part> ::= <regexp "[a-z]([-]?[a-z0-9])*">
+<port> ::= <number 1 to 65535>
+<path-part> ::= <regexp "[a-z0-9]([._-]?[a-z0-9])*">
+~~~
+
+## Discovery
+The discovery process involves resolving a name into namespace scoped metadata.
+The namespace metadata contains the full set of information needed to fetch and
+verify content associated with the namespace repository. The metadata includes
+list of registry api endpoints, the trust model, and search index. The discovery
+process should not be considered secure and therefore certificates retrieved as
+part of the discovery process should be verified before trusting.
+
+Discovery can be defined as...
+`<fully qualified name> -> scope([<registry api endpoint>, ...], <certificate>, <search index>, ...)`
+
+or in Go as...
+~~~go
+type NamespaceResolver interface {
+	Resolve(name string) Namespace
+}
+~~~
+
+### Default Method
+
+The first element of the namespace is extracted and used as the domain name for
+resolution. The domain name should be used unmodified.
+
+`<fully qualified name> -> <domain>/<path>`
+
+#### HTTPS
+The discovery related metadata will be fetched via HTTPS from the DNS resolved
+location using the remaining namespace path elements as the HTTP path in a GET
+request. A discovery request url would be in the format
+`https://<domain>/<name>?docker-discovery=1`
+
+For example, “example.com/foo/bar” would create a url
+“https://example.com/foo/bar?docker-discovery=1”
+
+##### HTML Response
+~~~html
+<meta name="docker-scope" content="example.com">
+<meta name=“docker-registry” content=“push,pull v2 https://registry.example.com/v2/”>
+<meta name=“docker-registry” content=“push,pull v1 https://registry.example.com/v1/”>
+<meta name=“docker-registry” content=“pull v2 https://registry.mirror.com/v2/”>
+<meta name=“docker-registry” content=“pull,notag v2 http://registry.mirror.com/v2/”>
+<meta name=“docker-index” content=“v1 https://search.mirror.com/v1/”>
+<meta name=“docker-trust” content=“https://trust.example.com/v2/”>
+<meta name=“docker-certificate” content=“https://trust.example.com/registry.cert”>
```

Comment:
circling back, I feel like the case for this is more along the lines of `docker-publisher-identity` and the content is the key or fingerprint to validate the images by. With the x.509 framework we could have better delegation for automated builders, etc. but this seems like an essential piece of namespaces